src/lsm/manifest.zig

```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             manifest_level.set_snapshot_max(snapshot, table);
             assert(table.snapshot_max == snapshot);
 
-            // Append update changes to the manifest log
+            // Append update changes to the manifest log.
             const log_level = @intCast(u7, level);
             manifest.manifest_log.insert(log_level, table);
         }
 
+        pub fn move_table(
+            manifest: *Manifest,
+            level_a: u8,
+            level_b: u8,
+            snapshot: u64,
+            table: *TableInfo,
+        ) void {
+            assert(level_b == level_a + 1);
+            assert(level_b < constants.lsm_levels);
+            assert(table.snapshot_max >= snapshot);
+
+            const manifest_level_a = &manifest.levels[level_a];
+            const manifest_level_b = &manifest.levels[level_b];
+            if (constants.verify) {
+                assert(manifest_level_a.contains(table));
+                assert(!manifest_level_b.contains(table));
+            }
+
+            // Insert the table into level b and append insert changes to the manifest log.
+            manifest_level_b.insert_table(manifest.node_pool, table);
+            manifest.manifest_log.insert(@intCast(u7, level_b), table);
+
+            // Update the table in level a without appending update changes to the manifest log.
+            // To move a table w.r.t manifest log, the previous level's table should not be removed.
+            // When replaying the log from open(), inserts are processed in LIFO order while
+            // duplicates are ignored. This means the table will only appear in level_b as moved.
+            manifest_level_a.set_snapshot_max(snapshot, table);
+            assert(table.snapshot_max == snapshot);
+
+            // Finally, the table address is recorded so that it will not be removed from the
+            // manifest log during `remove_invisible_tables`.
+            assert(manifest.moved_table_addresses[level_a] == null);
+            manifest.moved_table_addresses[level_a] = table.address;
```

Comment:
If I understand correctly, I think it would be better to use a different optional struct here, containing `level` and `address`. This ensures that only one move table can exist until `remove_invisible_tables()` is called.

Reply:
It needs multiple entries as we use the same `tree.manifest` to compact multiple `[lsm_levels/2]CompactionTable`s (+ `CompactionTableImmutable`) where each of them could do a `tree.maniefst.move_table()` under the same half-beat (`remove_invisible_tables()` is called when each half-beat ends).

Reply:
Thanks for the explanation @kprotty !

Would be fantastic to be able to remove this structure completely as @sentientwaffle suggested.